### PR TITLE
[RLlib] Fix connector registry

### PR DESCRIPTION
**Pull request by kouroshHakha**
Nice! 

image

## Why are these changes needed?
This fixes test_tune_restore with connectors enabled.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by kouroshHakha** [comment]
@gjoliver Please review, I have to disable connector in algorithm_config anyway, so I'll wait for your feedback and then flip the switch for merge.

File: rllib/connectors/__init__.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+from typing import Tuple, Any
+from ray.tune.registry import RLLIB_CONNECTOR, _global_registry
+
+from ray.util.annotations import PublicAPI
+from ray.rllib.connectors.connector import Connector, ConnectorContext
+
+
+@PublicAPI(stability="alpha")
+def register_connector(name: str, cls: Connector):
```
Comment (gjoliver) [inline rllib/connectors/__init__.py:9]: define these 2 in registry.py and just import here?

File: rllib/__init__.py [hunk 2]
```diff
@@ -58,6 +61,7 @@ def setup(self, config):
             register_trainable(alias, _see_contrib(alias))
 
 
+_register_all()
```
Comment (gjoliver) [inline rllib/__init__.py:64]: looks a bit suspicious. why didn't we have to register anything before this PR?

**Review by gjoliver — COMMENTED** [review]
looks good. a couple of quick questions.

**Review by gjoliver — COMMENTED** [review]
looks good. a couple of quick questions.

File: rllib/connectors/__init__.py (hunk 1 above)
Comment (kouroshHakha) [reply rllib/connectors/__init__.py:9]: Tried that, it game me circular dependency issues :( I solved circular dependency by this. Basically the problem is that pipleineconnectors use register_connector and get_connector. I cannot import them in the `registry.py` if registry.py also includes `register_connector`. The `register_connector` API has to be imported before both of these connector apis.

File: rllib/__init__.py (hunk 2 above)
Comment (kouroshHakha) [reply rllib/__init__.py:64]: We have explicitly called `_register_all()` in places where it was needed. for example in the tune_restore_test() we explicitly call this in the `tear_down()` function to reimport all the classes after the tune actor dies. For some reason that does not work for the connectors. We could have also called register_all when importing `algorithm.py` but I don't think that's the right place to call register_all. Basically the logic is that as soon as you import rllib all the namespaces with rllib should be recognizable globally.

**Comment by gjoliver** [comment]
Tests look good when I flip enable_connectors to True. Reverting the flag change now.
Screen Shot 2022-11-29 at 9 55 28 PM

**Review by gjoliver — APPROVED** [review]

File: rllib/connectors/registry.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+def _register_all_connectors():
+    """Force register all connectors again.
+
+    In case the cluster has been restarted, and the global registry
+    has to be rebuilt.
+    """
+    for module in ALL_CONNECTORS:
+        if module in sys.modules:
+            importlib.reload(sys.modules[module])
```
Comment (kouroshHakha) [inline rllib/connectors/registry.py:65]: nice. I learned something new :)

**Pull request merged**